subsys/net/ip/net_tc.c

```diff
@@ -65,10 +65,22 @@ static struct net_traffic_class rx_classes[NET_TC_RX_COUNT];
 enum net_verdict net_tc_submit_to_tx_queue(uint8_t tc, struct net_pkt *pkt)
 {
 #if NET_TC_TX_COUNT > 0
+#if NET_TC_TX_EFFECTIVE_COUNT > 1
+	uint8_t retry_cnt = 1;
+#endif
 	net_pkt_set_tx_stats_tick(pkt, k_cycle_get_32());
 
 #if NET_TC_TX_EFFECTIVE_COUNT > 1
+retry:
 	if (k_sem_take(&tx_classes[tc].fifo_slot, K_NO_WAIT) != 0) {
+		if (retry_cnt) {
+			retry_cnt--;
```

Comment:
I would organize a loop here instead of goto. And put at least a define for the retry count `#define NET_TC_RETRY_CNT 1`.
```suggestion
    for (uint8_t i = 0; i < NET_TC_RETRY_CNT; ++i) {
        if (k_sem_take(&tx_classes[tc].fifo_slot, K_NO_WAIT) == 0) {
            break;
        } else {
            /* Let thread with same priority run,
             * try to reduce dropping packets
             */
            k_yield();
        }
    }
```

Reply:
Hi @clamattia Then outside the loop, how to know it's NET_OK or NET_DROP?

Reply:
Did not think of that.
You could `goto drop` instead, which is a pattern seen sometimes. Or leave as is. Or @rlubos may have a suggestion.

```suggestion
{
#if NET_TC_TX_COUNT > 0
#if NET_TC_TX_EFFECTIVE_COUNT > 1
	uint8_t retry_cnt = NET_TC_RETRY_CNT;
#endif
	net_pkt_set_tx_stats_tick(pkt, k_cycle_get_32());

#if NET_TC_TX_EFFECTIVE_COUNT > 1
	while (k_sem_take(&tx_classes[tc].fifo_slot, K_NO_WAIT) != 0) {
		if (retry_cnt == 0) {
			goto drop;
		}
		retry_cnt--;
		/* Let thread with same priority run,
		 * try to reduce dropping packets
		 */
		k_yield();
	}
#endif

	k_fifo_put(&tx_classes[tc].fifo, pkt);
	return NET_OK;
#else
	ARG_UNUSED(tc);
	ARG_UNUSED(pkt);
drop:
	return NET_DROP;
#endif
}

```

Reply:
Thanks. please review again @jukkar @rlubos @clamattia